### PR TITLE
Pattern matching StringOrExpression

### PR DESCRIPTION
**Pull request by cjllanwarne**
One possible solution to the "`Poly`s are too much boilerplate" vs "`inl` and `inr` are too brittle" debate.

Also IMO much much easier to read the `match` `case`s

### CODE REVIEW

**Comment by Horneth** [comment]
That does look pretty good to me, enough to make me reverse my concerns about `unapply` at least in this case. 👍  as far as I"m concerned

File: cwl/src/main/scala/cwl/package.scala [hunk 1]
```diff
@@ -49,9 +49,16 @@ package object cwl extends TypeAliases {
...
 
   type StringOrExpression = Expression :+: String :+: CNil
-
   type Expression = ECMAScriptExpression :+: ECMAScriptFunction :+: CNil
 
+  object StringOrExpression {
+    object String {
+      def unapply(soe: StringOrExpression): Option[String] = soe.select[String]
+    }
+    object Expression {
```
Comment (danbills) [inline cwl/src/main/scala/cwl/package.scala]: can we take this one further and offer `ScriptExpression` and `ScriptFunction` that match to the nested `ECMAScriptExpression` and `ECMAScriptFunction`, respectively?

After all my talk of pattern matching being useful in nested coproducts I didn't even do it!

**Comment by mcovarr** [comment]
:+1: 🐦-proof

File: cwl/src/main/scala/cwl/package.scala (hunk 1 above)
Comment (cjllanwarne) [reply cwl/src/main/scala/cwl/package.scala]: I hadn't done initially since it wasn't being used (😄), but I don't see why not. Added.

**Pull request merged**